File: leetcode-codes/0803-cheapest-flights-within-k-stops/solution.py
from collections import deque
class Solution(object):
    def findCheapestPrice(self, n, flights, src, dst, k):
        
        q =deque()
        q.append((src,0,0))
        inf = 10000000000000
        dist=[inf]*n
        graph = [[] for i in range(n)]
        for f in flights:
            graph[f[0]].append((f[1],f[2]))
        dist[src]=0
        k+=1
        while q:
            cur,cost,cnt = q.popleft()
            ncnt =cnt+1
            for next,c in graph[cur]:
                ncost = cost+c
                if ncnt<=k and dist[next]>ncost:
                    dist[next]=ncost
                    q.append((next,ncost,ncnt))
        return dist[dst] if dist[dst]!=inf else -1

File: leetcode-codes/0803-cheapest-flights-within-k-stops/solution.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        vector<pair<int,int>> v[101];
        for(vector<int> &f :flights){
            v[f[0]].push_back({f[1],f[2]});
        }
        vector<int> dist(101,1e9);
        dist[src]=0;
        queue<pair<int,int>> q;
        q.push({src,0});
        int h=0;
        while(!q.empty()){
            int qsize= q.size();
            while(qsize--){
                int cur,sumcost;
                tie(cur,sumcost) = q.front();
                q.pop();
                for(int i =0;i<v[cur].size();i++){
                    int next = v[cur][i].first;
                    int ncost = sumcost+v[cur][i].second;
                    if(dist[next]>ncost){
                        dist[next]=ncost;
                        q.push({next,ncost});
                    }
                }
            }
            h++;
            if(h>k)
                break;
        }
        return dist[dst]==1e9?-1:dist[dst];
    }
};
